src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorPrimitives.netcore.cs

```diff
@@ -11785,6 +11785,17 @@ public static Vector512<T> Invoke(Vector512<T> t)
 
             public static Vector128<T> Invoke(Vector128<T> x)
             {
+#if NET9_0_OR_GREATER
+                if (typeof(T) == typeof(double))
+                {
+                    return Vector128.Log2(x.AsDouble()).As<double, T>();
+                }
+                else
+                {
+                    Debug.Assert(typeof(T) == typeof(float));
+                    return Vector128.Log2(x.AsSingle()).As<float, T>();
+                }
+#else
```

Comment:
Should we move these into the Log2Operator implementations instead? Each of its Invokes would then presumably become one-liners?

Either way is fine, but in some cases we compose one operator on top of another, so at least for those cases, doing it in the operator itself would serve double duty.

Reply:
I mean instead of this ifdef here, have it within Log2OperatorDouble.Invoke
https://github.com/dotnet/runtime/pull/96455/files/52d5a663edaf4b8e52fc56fb086bd80babf9a04b..4039954727505df784b86bb68f54c2d7d48ca2db#diff-9af9871a07c3bc8868a023467dd1e92b08eaee3e04fcc490126d89263678ed8eR11923

Reply:
I'm not sure that's "better".

That's going to:
* add an additional indirection for the .NET 9 case
* push the `ifdef` logic into (rather than around) the already complex implementation
* be dissimilar to how end users would need to compose these operations 
  * they can't access things like `LogOperatorDouble` and so have to go through `Log<double>`
* make the diff more complex when we move to an exclusively .NET 9+ world and only need to worry about dispatching to Vector64/128/256/512

I would also guess it's unnecessary for most of the APIs since we're either specializing (in which case we likely have a customized implementation anyways) or we're generic and we'd be calling the `Log2Operator<T>` version regardless.

Reply:
> add an additional indirection for the .NET 9 case

Just from a compilation persective, right? I'd hope they'd result in the same resulting asm.

> push the ifdef logic into (rather than around) the already complex implementation

From my perspective, that's a benefit, in that it means anyone reusing the implementation will get the best solution. If there's only one call site, it doesn't really matter; if there's ever multiple, then it matters (e.g. ExpOperator gets used in multiple places).

> be dissimilar to how end users would need to compose these operations

I'm not following. Can you elaborate?

In any event, like I said, either way is fine.

Reply:
> Just from a compilation persective, right? I'd hope they'd result in the same resulting asm.

It depends. There's a number of cases where the JIT will decide to emit a jump stub for trivial forwards rather than "inlining" the whole method or simply calling the actual target directly. I haven't checked in this exact case, but I've found it better to avoid those extra jumps where possible historically speaking.

> From my perspective, that's a benefit, in that it means anyone reusing the implementation will get the best solution. If there's only one call site, it doesn't really matter; if there's ever multiple, then it matters (e.g. ExpOperator gets used in multiple places).

I'm viewing `Log2OperatorSingle` as an implementation detail of `Log2Operator<T>` which itself is an implementation detail of the public `TensorPrimitives.Log2<T>`. I'd expect that most code just calls `TensorPrimitives.Log2` directly and iff we have any internal logic that really needs to reuse `Log2` (and I'm not aware of any currently), then it should use `Log2Operator<T>`. 

This also fits the longer term goal once we hit .NET 9+ where we no longer need to duplicate the actual implementation and we can simply call `Vector128.Log2()` directly and allows us to completely remove `Log2OperatorSingle/Double`. So any code using `TensorPrimitives.Log2<T>` doesn't need to be updated, where-as we'd have to update it if we did use `Log2OperatorSingle` directly.

This then also meshes with how users will end up building/composing their own, where they likewise have to use `TensorPrimitives.Log2<T>` and `Vector128.Log2<T>`

> In any event, like I said, either way is fine.

This is one I think I'll keep as is, because of the reasons above. I just think it fits more cleanly into the longer term code shape.